src/openrct2/world/Footpath.cpp

```diff
@@ -524,55 +524,50 @@ static void footpath_connect_corners(int32_t initialX, int32_t initialY, TileEle
         return;
 
     tileElement[0] = initialTileElement;
-    int32_t z = initialTileElement->base_height;
+    int32_t z = initialTileElement->GetBaseZ();
     for (int32_t initialDirection = 0; initialDirection < 4; initialDirection++)
     {
-        int32_t x = initialX;
-        int32_t y = initialY;
-        int32_t direction = initialDirection;
+        auto currentPos = footpathPos;
 
-        x += CoordsDirectionDelta[direction].x;
-        y += CoordsDirectionDelta[direction].y;
-        tileElement[1] = footpath_connect_corners_get_neighbour(x, y, z, (1 << direction_reverse(direction)));
+        tileElement[1] = footpath_connect_corners_get_neighbour({ currentPos, z }, (1 << direction_reverse(initialDirection)));
         if (tileElement[1] == nullptr)
             continue;
 
-        direction = (direction + 1) & 3;
-        x += CoordsDirectionDelta[direction].x;
-        y += CoordsDirectionDelta[direction].y;
-        tileElement[2] = footpath_connect_corners_get_neighbour(x, y, z, (1 << direction_reverse(direction)));
+        auto directionNext = (initialDirection + 1) & 3;
+        currentPos += CoordsDirectionDelta[directionNext];
+        tileElement[2] = footpath_connect_corners_get_neighbour({ currentPos, z }, (1 << direction_reverse(directionNext)));
         if (tileElement[2] == nullptr)
             continue;
 
-        direction = (direction + 1) & 3;
-        x += CoordsDirectionDelta[direction].x;
-        y += CoordsDirectionDelta[direction].y;
+        directionNext = (directionNext + 1) & 3;
+        currentPos += CoordsDirectionDelta[directionNext];
         // First check link to previous tile
-        tileElement[3] = footpath_connect_corners_get_neighbour(x, y, z, (1 << direction_reverse(direction)));
+
+        tileElement[3] = footpath_connect_corners_get_neighbour({ currentPos, z }, (1 << direction_reverse(directionNext)));
         if (tileElement[3] == nullptr)
             continue;
         // Second check link to initial tile
-        tileElement[3] = footpath_connect_corners_get_neighbour(x, y, z, (1 << ((direction + 1) & 3)));
+        tileElement[3] = footpath_connect_corners_get_neighbour({ currentPos, z }, (1 << ((directionNext + 1) & 3)));
         if (tileElement[3] == nullptr)
             continue;
 
-        direction = (direction + 1) & 3;
-        tileElement[3]->AsPath()->SetCorners(tileElement[3]->AsPath()->GetCorners() | (1 << (direction)));
-        map_invalidate_element({ x, y }, tileElement[3]);
+        directionNext = (directionNext + 1) & 3;
+        tileElement[3]->AsPath()->SetCorners(tileElement[3]->AsPath()->GetCorners() | (1 << (directionNext)));
+        map_invalidate_element(footpathPos + CoordsDirectionDelta[directionNext], tileElement[3]);
```

Comment:
This is too confusing to follow. Instead of recording just the tileElement in the array record the position as well using a CoordsXYE struct then you don't need to recalculate the position. (Note: direction will still need to be calculated as I can't at this moment follow if it stays the same)

Reply:
I stored a pair `std:;pair<PathElement*, CoordsXY>`, because I had to start storing `PathElement*` to comply with a suggestion above and they don't inherit from each other (thus we have to add some `reinterpret_cast`)